Refactor remote actor handling to Remote_Actors class

pfefferle opened the pull request:
This PR is for separation of concerns only, there is no new functionality added.

Moved all remote actor management logic from Actors to a new Remote_Actors class. Updated references throughout the codebase to use Remote_Actors for fetching, updating, error handling, and inbox management of remote actors. Deprecated old methods in Actors and replaced their usage in collections, handlers, models, REST controllers, and templates. Added tests for Remote_Actors functionality.

## Proposed changes:
- Extracts all remote actor-specific methods from `Actors` to new `Remote_Actors` class
- Updates references throughout the codebase to use `Remote_Actors` for remote actor operations
- Adds comprehensive test coverage for the new `Remote_Actors` functionality

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Click throught the Admin UI and check for errors.
* See tests.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

<!-- Add a changelog message here -->

</details>

pfefferle committed:
Refactor remote actor handling to Remote_Actors class

Moved all remote actor management logic from Actors to a new Remote_Actors class. Updated references throughout the codebase to use Remote_Actors for fetching, updating, error handling, and inbox management of remote actors. Deprecated old methods in Actors and replaced their usage in collections, handlers, models, REST controllers, and templates. Added tests for Remote_Actors functionality.

pfefferle committed:
Refactor to use Remote_Actors instead of Actors

Replaces usage of Activitypub\Collection\Actors with Activitypub\Collection\Remote_Actors in handler and test classes. This change improves clarity and consistency by using the more specific Remote_Actors class for remote actor operations.

pfefferle committed:
Refactor remote actor retrieval method calls

Replaces calls to get_remote_by_uri with get_by_uri in Actors and Remote_Actors classes for consistency and clarity in remote actor retrieval.

pfefferle committed:
Update import statements for actor functions

Moved the imports for is_actor and get_remote_metadata_by_actor from class-actors.php to class-remote-actors.php, and added object_to_uri import. This refactoring improves code organization and ensures functions are imported only where needed.

pfefferle committed:
Refactor tests to use Remote_Actors instead of Actors

Replaces usage of the Actors class with Remote_Actors in class-test-followers.php to improve clarity and maintainability. Also removes an obsolete test for clear_errors with invalid follower ID from class-test-actors.php.

pfefferle committed:
Remove unused public key variables from Test_Actors

Deleted several unused private key variables in the Test_Actors class to clean up the test file and improve maintainability.

pfefferle committed:
Remove unused imports from collection classes

Cleaned up unnecessary use statements in class-actors.php, class-blocked-actors.php, and class-followers.php to improve code clarity and maintainability.

pfefferle committed:
Remove unused constants from Actors class

Deleted POST_TYPE and CACHE_KEY_INBOXES constants from the Actors class as they are no longer used.

pfefferle committed:
Refactor Actors class: reorder and group key methods

Moved key management and identifier normalization methods to improve code organization and readability. No functional changes; methods were relocated within the class for better logical grouping.

pfefferle committed:
Update deprecated version in get_remote_key docblock

Changed the @deprecated version and deprecation notice in Signature::get_remote_key from '7.0.0' to 'unreleased' to reflect that the function is not yet deprecated in a released version.

pfefferle committed:
Replace Actors::POST_TYPE with Remote_Actors::POST_TYPE

Updated all references to Actors::POST_TYPE to use Remote_Actors::POST_TYPE across core, collections, migrations, and tests. This change improves clarity and consistency in handling remote actor post types throughout the codebase.

pfefferle committed:
Update EC public key in remote actors test

THE AI SIMPLY CHANGED THE KEY 🤯

pfefferle committed:
Import strip_fragment_from_url in Remote Actors class

Added the import for the strip_fragment_from_url function in class-remote-actors.php to enable its usage within the Remote Actors collection class.

pfefferle committed:
Fix namespace usage for strip_fragment_from_url

Replaces the local function import with a fully qualified namespace call to strip_fragment_from_url in get_public_key. This ensures the correct function is used and resolves potential namespace issues.

pfefferle committed:
Merge branch 'trunk' into change/add-remote-actors-collection

obenland commented on includes/collection/class-remote-actors.php:
```diff
@@ -0,0 +1,488 @@
+			),
+		);
+	}
+
+	/**
+	 * Normalize actor identifier to a URI.
+	 *
+	 * Handles webfinger addresses, URLs without schemes, objects, and arrays.
+	 *
+	 * @param string|object|array $actor Actor URI, webfinger address, actor object, or array.
+	 * @return string|null Normalized actor URI or null if unable to resolve.
+	 */
+	public static function normalize_identifier( $actor ) {
+		$actor = object_to_uri( $actor );
+		if ( ! is_string( $actor ) ) {
+			return null;
+		}
+
+		$actor = \trim( $actor, '@' );
+
+		// If it's an email-like webfinger address, resolve it.
+		if ( \filter_var( $actor, FILTER_VALIDATE_EMAIL ) ) {
+			$resolved = \Activitypub\Webfinger::resolve( $actor );
+			return \is_wp_error( $resolved ) ? null : object_to_uri( $resolved );
+		}
+
+		// If it's a URL without scheme, add https://.
+		if ( empty( \wp_parse_url( $actor, PHP_URL_SCHEME ) ) ) {
+			$actor = \esc_url_raw( 'https://' . \ltrim( $actor, '/' ) );
+		}
+
+		return $actor;
+	}
```
```suggestion
```
This doesn't appear to be used, should we just stick with the one in `Actors`?

pfefferle replied:
Good catch, but it belongs to to the Remote_Actors class, so I removed it from Actors ;)

obenland commented on includes/collection/class-actors.php:
```diff
@@ -1060,4 +620,217 @@ protected static function check_legacy_key_pair( $user_id ) {
+	/**
+	 * Returns all Inboxes for all known remote Actors.
+	 *
+	 * @deprecated unreleased Use {@see Activitypub\Collection\Remote_Actors::get_inboxes()}
```
```suggestion
	 * @deprecated unreleased Use {@see Remote_Actors::get_inboxes()}
```
Not sure if it's just a quirk of my IDE, but `Activitypub\Collection\Remote_Actors` without the leading backslash (`\Activitypub`) creates `Undefined namespace 'Collection'` notices. 

Since this annotation is meant for IDE traversal and `Actor` and `Remote_Actor` are in the same namespace we can just remove it from this and following annotations.

pfefferle committed:
class-actors.php aktualisieren

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Merge branch 'trunk' into change/add-remote-actors-collection

pfefferle committed:
Refactor actor identifier normalization to Remote_Actors

Moved the normalize_identifier method from Actors to Remote_Actors and updated all references accordingly. The original method in Actors is now deprecated and delegates to Remote_Actors. This improves separation of concerns and clarifies responsibility for actor normalization logic.

obenland committed:
Update remaining @deprecated annotations

obenland committed:
Update namespace imports

obenland committed:
Update deprecation notices

obenland reviewed (approved):
🤞
